### PR TITLE
fix when you create-amplify absolute path, raised error

### PR DESCRIPTION
**Pull request by magisystem0408**
## Problem

**Issue number, if available: **
fixes: #827

Related pull requests: https://github.com/aws-amplify/amplify-backend/pull/857

## Changes

**Corresponding docs PR, if applicable:**

when you do create-amplify absolute path (ex. '/user/test'), raised error and log.

## Checklist

- [x] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [x] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [x] If this PR requires a docs update, I have linked to that docs PR above.
- [x] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

**Review by sobolk — DISMISSED** [review]

**Comment by sobolk** [comment]
Note to other reviewers: this change was discussed in https://github.com/aws-amplify/amplify-backend/pull/857 and got moved to new PR.

File: packages/create-amplify/src/get_project_root.ts [hunk 1]
```diff
@@ -37,7 +37,17 @@ export const getProjectRoot = async () => {
       LogLevel.DEBUG
     );
     printer.log(`Creating directory ${projectRoot}`, LogLevel.DEBUG);
-    await fsp.mkdir(projectRoot, { recursive: true });
+    try {
+      await fsp.mkdir(projectRoot, { recursive: true });
+    } catch (err) {
+      if (path.isAbsolute(projectRoot)) {
```
Comment (Amplifiyer) [inline packages/create-amplify/src/get_project_root.ts:43]: In what conditions would a `projectRoot` not be an absolute path? 

https://github.com/aws-amplify/amplify-backend/pull/1071/files#diff-17a8da21bcc08aa19a0287db1113f604841ad4119d227505260fdc18ac081181R26-R28

Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.ts:43]: I can't think of a case at the moment where projectRoot not be an absolute path, but I put it in explicitly.

Is it necessary?

**Review by edwardfoyle — DISMISSED** [review]

File: packages/create-amplify/src/get_project_root.ts (hunk 1 above)
Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.ts]: Thank you feedback !!! @edwardfoyle @Amplifiyer 

I try to do this task!
- [ ] remove condition
- [ ] add `AmplifyError`

Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.ts]: @Amplifiyer @edwardfoyle 

I came up with three error pattern.
should select plan3 right?

```
// plan1: use printer
try {
    await fsp.mkdir(projectRoot, {recursive: true});
} catch (err) {
    printer.log(
        `Failed to create directory at ${projectRoot}. Ensure this is the correct path and you have write permissions to this location.`,
        LogLevel.ERROR
    );
    throw err;
}
------------------
// plan2: direct throw new Error
try {
    await fsp.mkdir(projectRoot, {recursive: true});
} catch (err) {
    throw new Error(`Failed to create directory at ${projectRoot}. Ensure this is the correct path and you have write permissions to this location.`);
}
------------------
// plan3: use AmplifyUserError
import { AmplifyUserError } from '@aws-amplify/platform-core';

throw new AmplifyUserError('MultipleSingletonResourcesError', {
    message:
        `Failed to create directory at ${projectRoot}. Ensure this is the correct path and you have write permissions to this location.`,
    resolution: 'Change relative path or Absolute path under root',
});
```

Comment (Amplifiyer) [reply packages/create-amplify/src/get_project_root.ts]: Yes plan3 is correct. Make the error name something like `ProjectDirectoryCreateError` and change the message and resolution to something like 

```ts
try {
    await fsp.mkdir(projectRoot, {recursive: true});
} catch (err) {
    throw new AmplifyUserError('MultipleSingletonResourcesError', {
        message:
            `Failed to create project directory`,
        resolution: 'Ensure that ${projectRoot} is the correct path and you have write permissions to this location.',
    });
}
```

Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.ts]: https://github.com/aws-amplify/amplify-backend/pull/1071/commits/4e070def49e77f1560468835a90fbc837cfe1358

**Comment by magisystem0408** [comment]
@Amplifiyer 

I modified the test code and AmplifyError with reference to the following.

https://github.com/aws-amplify/amplify-backend/blob/8d9a7a4c3033c474b0fc78379cdd4c1854d890ce/packages/platform-core/src/errors/amplify_error.test.ts#L20-L27

File: packages/create-amplify/src/get_project_root.test.ts [hunk 2]
```diff
@@ -72,6 +74,48 @@ void describe('getProjectRoot', () => {
...
+          expectedError.classification
+        );
+        assert.deepStrictEqual(actual?.message, expectedError.message);
+        assert.deepStrictEqual(actual?.details, expectedError.details);
+        assert.deepStrictEqual(actual?.cause?.name, expectedError.cause?.name);
+        assert.deepStrictEqual(
+          actual?.cause?.message,
+          expectedError.cause?.message
+        );
+        return true;
```
Comment (Amplifiyer) [inline packages/create-amplify/src/get_project_root.test.ts:109]: You don't need all this. Just assert on the `err` that the `name`, `message` and `resolution` is correct.

Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.test.ts]: try change!

Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.test.ts]: @Amplifiyer 
thank you feedback!!!!!
I try change test case

File: packages/create-amplify/src/get_project_root.test.ts [hunk 3]
```diff
@@ -72,6 +74,39 @@ void describe('getProjectRoot', () => {
...
+        assert.deepStrictEqual(actual?.message, expectedError.message);
+        assert.deepStrictEqual(actual?.resolution, expectedError.resolution);
+        return true;
+      }
+    );
+    assert.equal(fsMkDirSyncMock.mock.callCount(), 1);
+    assert.equal(
+      fsMkDirSyncMock.mock.calls[0].arguments[0],
+      path.resolve(userInput)
+    );
```
Comment (Amplifiyer) [inline packages/create-amplify/src/get_project_root.test.ts:107]: something like below
```suggestion
    await assert.throws(
      () => getProjectRoot(),
      new AmplifyUserError('ProjectDirectoryCreateError', {
        message: `Failed to create project directory`,
        resolution: `Ensure that ${userInput} is the correct path and you have write permissions to this location.`,
      });
    );
    assert.equal(fsMkDirSyncMock.mock.callCount(), 1);
    assert.equal(
      fsMkDirSyncMock.mock.calls[0].arguments[0],
      path.resolve(userInput)
    );
```

Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.test.ts]: @Amplifiyer 

I've made a few modifications.
at [d7226b6](https://github.com/aws-amplify/amplify-backend/pull/1071/commits/d7226b6a974c18e0362af236f848bc51c0e1bc70)

Comment (magisystem0408) [reply packages/create-amplify/src/get_project_root.test.ts]: I came up with idea.
```
const expectedError = new AmplifyUserError('ProjectDirectoryCreateError', {
      message: `Failed to create project directory`,
      resolution: `Ensure that ${
        userInput
      } is the correct path and you have write permissions to this location.`,
    });
    fsMkDirSyncMock.mock.mockImplementationOnce(() => {
      throw expectedError;
    });

    await assert.rejects(() => getProjectRoot(), {
      name: expectedError.name,
      message: expectedError.message,
    });
```

**Review by edwardfoyle — APPROVED** [review]

**Review by Amplifiyer — APPROVED** [review]

**Pull request merged**